### ISSUE TITLE
3.3.5: 3 Barracks builds delay hero 5+ minutes (not supply-blocked)

### ISSUE THREAD

**Issue by Slayer95** [SMUnlimited/AMAI#375]
Observed in Last Refuge with Humans, and Undead (Crypt).

Hero comes up after 6 min mark, and it never leaves its base.

[AMAI-Crypt-Broken-Hero.zip](https://github.com/user-attachments/files/17085946/AMAI-Crypt-Broken-Hero.zip)

**Comment by jzy-chitong56** [comment]
```
SleepInCombatAM
    if (u != null and (not CaptainInCombat(true) or CaptainRetreating() or CaptainIsHome())) then
```
maybe 
   ` if (u != null and not (CaptainInCombat(true) and CaptainRetreating() and CaptainIsHome())) then`


There is indeed feedback that the 3.3.5a human race only goes out when they reach 80 food points

**Comment by SMUnlimited** [comment]
> Observed in Last Refuge with Humans, and Undead (Crypt).
> 
> Hero comes up after 6 min mark, and it never leaves its base.
> 
> [AMAI-Crypt-Broken-Hero.zip](https://github.com/user-attachments/files/17085946/AMAI-Crypt-Broken-Hero.zip)

Can't get replay to run, was this undead or human or both? There was an issue with undead fixed in 3.3.5a which made it do nothing for ~5ish minutes.

**Comment by Slayer95** [comment]
> > Observed in Last Refuge with Humans, and Undead (Crypt).
> > Hero comes up after 6 min mark, and it never leaves its base.
> > [AMAI-Crypt-Broken-Hero.zip](https://github.com/user-attachments/files/17085946/AMAI-Crypt-Broken-Hero.zip)
> 
> Can't get replay to run, was this undead or human or both? There was an issue with undead fixed in 3.3.5a which made it do nothing for ~5ish minutes.

That replay is 1.27b Undead, but I have also seen this with Human just yesterday.

**Comment by Slayer95** [comment]
And you may also find the Undead replay at https://wc3stats.com/games/584008/events

If you can't see either replay, the wc3stats site shows the build orders. Just go to the "Events" tab.

**Comment by SMUnlimited** [comment]
Aha I only test reforged regularly so probably an issue with the classic TFT scripts, why i don't see these issues and why your replay not loading. If you know which strategy undead/human was running or it says in the replays I can work out why its delaying fairly easy.

**Comment by Slayer95** [comment]
Human went ``I'm massing Spell Casters``.

![image](https://github.com/user-attachments/assets/92bd5a1e-e858-4c77-b849-d89baba73850)


Undead went full Gargoyles as per the following screenshot (but I don't remember the AMAI strategy):

![image](https://github.com/user-attachments/assets/1b2ed5be-d5e0-4027-9d72-2aaa2af725fc)

**Comment by jzy-chitong56** [comment]
1.27b is almost impossible to find resources web, even though it remains in a lower version in China. Here, 1.27a is used. 
I suggest you switch to a different version

**Comment by jzy-chitong56** [comment]
@SMUnlimited 

If you are certain that it is an individual Strategy exception, why not directly compare the two versions of the policy file `Strategy.txt`

Please handle this issue along the way
https://github.com/SMUnlimited/AMAI/issues/356

**Comment by SMUnlimited** [comment]
@Slayer95 Try latest commit which is a quick fix, just some obvious places were the hero priority was just too low or on the boundary to other units.

**Comment by Slayer95** [comment]
Oh! I am very sorry. I should have realized that it was 100% relevant for this issue. I mentioned that the replay was 1.27b so that you could load it. But the map is actually running REFORGED strategies and StandardUnits (as well as the relevant ``w3u`` unit data).

However, I am very confident that my merging was done properly. So, can I trouble you to at least have a look at the ``MassSp`` REFORGED strategy? According to your commit, any issue should lie in the priorities, but this ``SetBuildReact`` function is a bit obscure. However, if you already did or aren't willing to, I'd totally understand if you closed this issue.

Once again, apologies for the incomplete report.

**Comment by jzy-chitong56** [comment]
> @Slayer95 Try latest commit which is a quick fix, just some obvious places where the hero priority was just too low or on the boundary to other units.

I don't understand the meaning, I think this change is ineffective. If you look at other strategies, you will find that they are generally low, and increasing certain priorities alone should not solve the problem

I suggest restoring this change and investigating the real reason



@Slayer95 

> SetBuildReact
This feature is also based on an official function modification, originally intended to dynamically adjust the training quantity of two units based on food, but there is a fallback quantity. Do you think it is not working properly?
_Basically, I'm just following along_

**Comment by SMUnlimited** [comment]
> Oh! I am very sorry. I should have realized that it was 100% relevant for this issue. I mentioned that the replay was 1.27b so that you could load it. But the map is actually running REFORGED strategies (i.e. BuildSequence, etc.) and StandardUnits (as well as the relevant `w3u` unit data).
> 
> However, I am very confident that my merging was done properly. So, can I trouble you to at least have a look at the `MassSp` REFORGED strategy? According to your commit, any issue should lie in the priorities, but this `SetBuildReact` function is a bit obscure. However, if you already did or aren't willing to, I'd totally understand if you closed this issue.
> 
> Once again, apologies for the incomplete report.

Oh interesting. Well TFT was a genuine bug that was fixed in reforged scripts and I could reproduce that issue.
I started ~50 games on REFORGED on your above map and the issue didn't occur. Only slight delay was when buying a tavern hero as normal.  Don't know how often it happens for you?

If your using reforged scripts some thoughts:
- Its a bug only with 1.27.x and the REFORGED scripts, but this isn't technically supported because there may be a incompatible tech tree changes blocking the AI.
- Its a bug only with 1.27.x regardless of the TFT or REFORGED scripts, possible but unlikely as we compile against an old API. 
- Its just so rare I just been super lucky. 

If you can reproduce with TFT scripts and 1.27.x, or REFORGED scripts and 1.36.x it will help me know which problem it might be.

**Comment by Slayer95** [comment]
Thanks for looking in to this.

Indeed, last night I was running tests, and even when Human went Mass Sp, I couldn't reproduce this issue. But I was surprised that it originally happened twice in the same day, when I never before had seen this.

So, going forward, I intend to (1) gather more W3MMD data, and/or (2) build the map with only suspect AMAI strategies enabled, to get a clearer picture. I may also look into testing with 1.36x, once I have figured out (2).

**Comment by Slayer95** [comment]
> SetBuildReact
This feature is also based on an official function modification, originally intended to dynamically adjust the training quantity of two units based on food, but there is a fallback quantity. Do you think it is not working properly?

I am sorry, I am unable to judge this without me further investigating the innards of AMAI.

**Comment by jzy-chitong56** [comment]
@Slayer95 
You'd better play the video yourself, record it and upload it to the video platform, so there won't be any problem of not being able to play it
_Of course, if you share on YouTube, China still cannot directly play it. Ahhhhhhhhhh. emmm_

**Comment by Slayer95** [comment]
> Only slight delay was when buying a tavern hero as normal

This rang a bell.

I am 90% sure that this delay happened because I have recently been playing more of the non-melee modes of the map. In these game modes, the tavern is replaced by another custom units. So whenever AI wants to get a T1 tavern hero, they fail to do so. Instead, they get their T2 hero as their first hero. This is consistent with the timing of T2 (Tech and Expansion) and first hero Build Order shown in the replay analytics above.

Therefore, I probably need to tinker myself with AMAI and/or my map, so they armonize well in these custom modes.

If I get evidence to disprove this hypothesis, I will try to save a 1.27a replay for you, @jzy-chitong56 .

Meanwhile, I think this issue can be closed.

Thanks to you both for your attention.
